user_badges.module

```diff
@@ -37,10 +37,15 @@ function user_badges_theme() {
  */
 function user_badges_user_presave(Drupal\Core\Entity\EntityInterface $entity) {
   $badges = array();
+  $entity_manager = \Drupal::entityTypeManager();
   /** @var \Drupal\user\Entity\User $entity */
   $roles = $entity->getRoles();
   foreach ($roles AS $role) {
```

Comment:
Instead of this `foreach`, do `loadByProperties(array('role_id' => $roles)`.  `EntityStorageBase::buildPropertyQuery` will use the array just fine and entity query keys by badge id (because we are not revisionable, otherwise it would key by revision id and you'd need to array_unique the results) so badge ids can't repeat.